quickevent/app/quickevent/plugins/CardReader/src/services/qropunch.cpp

```diff
@@ -0,0 +1,216 @@
+#include "qropunch.h"
+#include "qropunchwidget.h"
+
+#include "../cardreaderplugin.h"
+
+#include <quickevent/core/si/checkedcard.h>
+#include <quickevent/core/codedef.h>
+
+#include <qf/qmlwidgets/framework/mainwindow.h>
+
+#include <qf/core/log.h>
+
+#include <QTcpSocket>
+#include <QTimer>
+#include <regex>
+
+using qf::qmlwidgets::framework::getPlugin;
+
+namespace CardReader {
+namespace services {
+
+QrOPunch::QrOPunch(QObject *parent)
+	: Super(QrOPunch::serviceName(), parent)
+{
+	auto onSettingsChanged = [this] {
+		if (m_tcpServer) {
+			init();
+		}
+	};
+	connect(this, &QrOPunch::settingsChanged, this, onSettingsChanged, Qt::QueuedConnection);
+}
+
+void QrOPunch::run()
+{
+	init();
+	Super::run();
+}
+
+void QrOPunch::stop()
+{
+	Super::stop();
+	m_tcpClients.clear();
+	delete m_tcpServer, m_tcpServer = nullptr;
+}
+
+QString QrOPunch::serviceName()
+{
+	return QStringLiteral("QrOPunch");
+}
+
+qf::qmlwidgets::framework::DialogWidget *QrOPunch::createDetailWidget()
+{
+	auto *w = new QrOPunchWidget();
+	return w;
+}
+
+void QrOPunch::init()
+{
+	QrOPunchSettings ss = settings();
+	int port = ss.tcpListenPort();
+
+	m_tcpClients.clear();
+	if (m_tcpServer) {
+		delete m_tcpServer;
+	}
+	m_tcpServer = new QTcpServer(this);
+	connect(m_tcpServer, &QTcpServer::newConnection, this, &QrOPunch::onNewConnection);
+
+	if (!m_tcpServer->listen(QHostAddress::Any, port)) {
+		qfError() << "Server could not start listening port " << port << ": " << m_tcpServer->errorString();
+	} else {
+		qfInfo() << "TCP server started listening port " << port;
+	}
+	if (ss.isWriteLogFile()) {
+		m_logFileFs.close();
+        m_logFileFs.open(ss.logFileName().toLocal8Bit(), std::ios_base::app);
+        if (!m_logFileFs) {
+			qfError() << "Failed to open file for logging: " << ss.logFileName();
+		} else {
+			qfInfo() << "Opened file for logging: " << ss.logFileName();
+		}
+	} else {
+		m_logFileFs.close();
+	}
+}
+
+void QrOPunch::onNewConnection()
+{
+	while (m_tcpServer->hasPendingConnections())
+	{
+		QTcpSocket *socket = m_tcpServer->nextPendingConnection();
+
+		connect(socket, &QTcpSocket::readyRead, this, &QrOPunch::onReadyRead);
+		connect(socket, &QTcpSocket::disconnected, this, [this] {
+			QTcpSocket *socket = qobject_cast<QTcpSocket*>(sender());
+			m_tcpClients.remove(socket->socketDescriptor());
+			socket->deleteLater();
+		});
+		m_tcpClients.insert(socket->socketDescriptor(), {});
+	}
+}
+
+void QrOPunch::onReadyRead()
+{
+	QTcpSocket *socket = qobject_cast<QTcpSocket*>(sender());
+	auto sendError = [socket](const char *error) {
+		QTimer::singleShot(3000, socket, [socket] { socket->close(); });
+		socket->write("HTTP/1.1 ");
+		socket->write(error);
+		socket->write("\r\n\r\n");
+		socket->flush();
+		socket->connect(socket, &QTcpSocket::bytesWritten, [socket](qint64) {
+			if (!socket->bytesToWrite())
+				socket->close();
+		});
+	};
+	auto it = m_tcpClients.find(socket->socketDescriptor());
```

Comment:
```c++
if(auto it = m_tcpClients.find(socket->socketDescriptor()); it != m_tcpClients.end()) {
	QByteArray &buffer = it.value();
	while (socket->bytesAvailable() > 0)
	{
		buffer.append(socket->readAll());
```
Please check validity of iterator, even if it should be always valid or use `assert`, if you want. We need some clue in case of app crash.

Reply:
I agree, let me put an assert here because the socket is always expected to be in the map.